contracts/random/RandomSeedProviderRequestQueue.sol

```diff
@@ -0,0 +1,151 @@
+// Copyright (c) Immutable Pty Ltd 2018 - 2023
+// SPDX-License-Identifier: Apache 2
+pragma solidity 0.8.19;
+
+/**
+ * @notice Queue to be used with RandomSeedProvider
+ * @dev    The contract is upgradeable.
+ */
+contract RandomSeedProviderRequestQueue {
```

Comment:
I might be misunderstanding something, but the implementation of this contract feels a bit inefficient. 
The functions that remove items from the queue seem to do so based on block number (e.g. `dequeueBlockNumber()`) and not strictly based on their order in the queue? If so, it feels like the implementation can be better optimised for this

Reply:
I will add more information in the contract level comment. This is to prevent an attack: attacker generates random for many block numbers, but doesn't fulfil. If the queue only extracted one block number at a time, then the user would have to read each block number and do an sstore, to store the block hash for that block number, and keep processing until the hit their own block number. At 20,000 gas per sstore, and possibly 250 block numbers to process, that would be 5,000,000 gas that an attacker could force a normal user to have to use.